include/proj/util.hh

```diff
@@ -0,0 +1,396 @@
+using LocalNameNNPtr = util::nn<LocalNamePtr>;
+
+class NameSpace;
+using NameSpacePtr = std::shared_ptr<NameSpace>;
+using NameSpaceNNPtr = util::nn<NameSpacePtr>;
+
+class GenericName;
+using GenericNamePtr = std::shared_ptr<GenericName>;
+using GenericNameNNPtr = util::nn<GenericNamePtr>;
+
+// ---------------------------------------------------------------------------
+
+class GenericName : public BaseObject {
+  public:
+    PROJ_DLL GenericName(const GenericName &other);
+    PROJ_DLL GenericName &operator=(const GenericName &other) = delete;
+    PROJ_DLL virtual ~GenericName();
+
+    PROJ_DLL virtual const NameSpacePtr scope() const = 0;
+    PROJ_DLL virtual std::string toString() const = 0;
+    PROJ_DLL virtual GenericNameNNPtr toFullyQualifiedName() const = 0;
+
+  protected:
+    GenericName();
+
+  private:
+    PROJ_OPAQUE_PRIVATE_DATA
+};
+
+// ---------------------------------------------------------------------------
+
+class NameSpace {
+  public:
+    PROJ_DLL NameSpace(const NameSpace &other);
+    PROJ_DLL NameSpace &operator=(const NameSpace &other) = delete;
+    PROJ_DLL ~NameSpace();
+
+    PROJ_DLL bool isGlobal() const;
+    PROJ_DLL const GenericNamePtr &name() const;
+
+  protected:
+    friend class NameFactory;
+    friend class LocalName;
+    explicit NameSpace(const GenericNamePtr &name);
+    NameSpaceNNPtr getGlobalFromThis() const;
+    const std::string &separator() const;
+    static const NameSpaceNNPtr GLOBAL;
+    INLINED_MAKE_SHARED
+
+  private:
+    PROJ_OPAQUE_PRIVATE_DATA
+
+    static NameSpaceNNPtr createGLOBAL();
+};
+
+// ---------------------------------------------------------------------------
+
+class LocalName : public GenericName {
+  public:
+    PROJ_DLL LocalName(const LocalName &other);
+    PROJ_DLL LocalName &operator=(const LocalName &other) = delete;
+    ~LocalName();
+
+    PROJ_DLL const NameSpacePtr scope() const override;
+    PROJ_DLL std::string toString() const override;
+    PROJ_DLL GenericNameNNPtr toFullyQualifiedName() const override;
+
+  protected:
+    friend class NameFactory;
+    friend class NameSpace;
+    explicit LocalName(const std::string &nameIn);
+    LocalName(const NameSpacePtr &ns, const std::string &name);
+    INLINED_MAKE_SHARED
+
+  private:
+    PROJ_OPAQUE_PRIVATE_DATA
+};
+
+// ---------------------------------------------------------------------------
+
+class NameFactory {
+  public:
+    PROJ_DLL static NameSpaceNNPtr
+    createNameSpace(const GenericNameNNPtr &name,
+                    const PropertyMap &properties);
+    PROJ_DLL static LocalNameNNPtr createLocalName(const NameSpacePtr &scope,
+                                                   const std::string &name);
+    PROJ_DLL static GenericNameNNPtr
+    createGenericName(const NameSpacePtr &scope,
+                      const std::vector<std::string> &parsedNames);
+};
+
+// ---------------------------------------------------------------------------
+
+class CodeList {
```

Comment:
The `CodeList` feels like it could be `using CodeList = std::string` :)

Reply:
Not completely. This is the base class to create enumeration of strings. See AxisDirection in coordinatesystem.hpp